File: Compiler/Haskell/Syntax/src/Visual/XSharp/Diagnostic/Protocol.hs
-- SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
-- SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.1

module Visual.XSharp.Diagnostic.Protocol
    ( DiagnosticDocument (..)
    , DiagnosticRecord (..)
    , ProtocolSeverity (..)
    , DiagnosticArgument (..)
    , DiagnosticLocation (..)
    , DiagnosticRelatedLocation (..)
    , DiagnosticFix (..)
    , DiagnosticTextEdit (..)
    , DiagnosticProtocolLimits (..)
    , DiagnosticProtocolError (..)
    , defaultDiagnosticProtocolLimits
    , diagnosticDocument
    , encodeDiagnosticDocument
    , decodeDiagnosticDocument
    ) where

import Control.Monad (replicateM, unless, when)
import Data.Bits (shiftL, (.|.))
import Data.ByteString qualified as ByteString
import Data.ByteString.Builder qualified as Builder
import Data.ByteString.Lazy qualified as LazyByteString
import Data.Char (isAsciiUpper, isDigit, ord)
import Data.List (nub)
import Data.Word (Word16, Word32, Word8)
import Visual.XSharp.AST (SourcePosition (..), SourceSpan (..))
import Visual.XSharp.Diagnostic

protocolVersion :: Word16
protocolVersion = 1

magic :: [Word8]
magic = map (fromIntegral . ord) "VXDG"

data DiagnosticArgument = DiagnosticArgument
    { argumentName :: String
    , argumentValue :: String
    }
    deriving (Eq, Ord, Read, Show)

-- The frontend currently emits errors and warnings, while native stages may
-- also emit information and hints. Keeping the wire catalog distinct prevents
-- the syntax diagnostic API from acquiring severities it does not yet use.
data ProtocolSeverity
    = ProtocolError
    | ProtocolWarning
    | ProtocolInformation
    | ProtocolHint
    deriving (Bounded, Enum, Eq, Ord, Read, Show)

-- Protocol positions are zero-based UTF-16-independent source coordinates.
-- The compiler currently counts Unicode scalar columns; consumers must not
-- reinterpret them as byte offsets or JVM UTF-16 offsets.
data DiagnosticLocation = DiagnosticLocation
    { locationSource :: FilePath
    , locationStartLine :: Word32
    , locationStartColumn :: Word32
    , locationEndLine :: Word32
    , locationEndColumn :: Word32
    }
    deriving (Eq, Ord, Read, Show)

data DiagnosticRelatedLocation = DiagnosticRelatedLocation
    { relatedLocation :: DiagnosticLocation
    , relatedMessage :: String
    }
    deriving (Eq, Ord, Read, Show)

data DiagnosticTextEdit = DiagnosticTextEdit
    { editLocation :: DiagnosticLocation
    , editReplacement :: String
    }
    deriving (Eq, Ord, Read, Show)

data DiagnosticFix = DiagnosticFix
    { fixTitle :: String
    , fixEdits :: [DiagnosticTextEdit]
    }
    deriving (Eq, Ord, Read, Show)

data DiagnosticRecord = DiagnosticRecord
    { recordStage :: DiagnosticStage
    , recordSeverity :: ProtocolSeverity
    , recordCode :: String
    , recordMessage :: String
    , recordArguments :: [DiagnosticArgument]
    , recordPrimary :: Maybe DiagnosticLocation
    , recordRelated :: [DiagnosticRelatedLocation]
    , recordFixes :: [DiagnosticFix]
    }
    deriving (Eq, Ord, Read, Show)

newtype DiagnosticDocument = DiagnosticDocument
    { documentRecords :: [DiagnosticRecord]
    }
    deriving (Eq, Ord, Read, Show)

data DiagnosticProtocolLimits = DiagnosticProtocolLimits
    { maximumWireBytes :: Int
    , maximumRecords :: Int
    , maximumTextScalars :: Int
    , maximumArguments :: Int
    , maximumRelatedLocations :: Int
    , maximumFixes :: Int
    , maximumEditsPerFix :: Int
    }
    deriving (Eq, Ord, Read, Show)

defaultDiagnosticProtocolLimits :: DiagnosticProtocolLimits
defaultDiagnosticProtocolLimits =
    DiagnosticProtocolLimits
        { maximumWireBytes = 16 * 1024 * 1024
        , maximumRecords = 65535
        , maximumTextScalars = 1024 * 1024
        , maximumArguments = 256
        , maximumRelatedLocations = 256
        , maximumFixes = 128
        , maximumEditsPerFix = 4096
        }

data DiagnosticProtocolError = DiagnosticProtocolError
    { protocolErrorOffset :: Int
    , protocolErrorContext :: String
    , protocolErrorMessage :: String
    }
    deriving (Eq, Ord, Read, Show)

diagnosticDocument :: [Diagnostic] -> Either DiagnosticProtocolError DiagnosticDocument
diagnosticDocument diagnostics = DiagnosticDocument <$> traverse convert diagnostics
    where
        convert value = do
            primary <- traverse sourceLocation (diagnosticSpan value)
            pure
                DiagnosticRecord
                    { recordStage = diagnosticStage value
                    , recordSeverity = protocolSeverity (diagnosticSeverity value)
                    , recordCode = diagnosticCode value
                    , recordMessage = diagnosticMessage value
                    , recordArguments = []
                    , recordPrimary = primary
                    , recordRelated = []
                    , recordFixes = []
                    }

sourceLocation :: SourceSpan -> Either DiagnosticProtocolError DiagnosticLocation
sourceLocation spanValue = do
    startLine <- sourceCoordinate "source start line" (sourceLine (sourceStart spanValue))
    startColumn <- sourceCoordinate "source start column" (sourceColumn (sourceStart spanValue))
    endLine <- sourceCoordinate "source end line" (sourceLine (sourceEnd spanValue))
    endColumn <- sourceCoordinate "source end column" (sourceColumn (sourceEnd spanValue))
    pure
        DiagnosticLocation
            { locationSource = sourceFile spanValue
            , locationStartLine = startLine
            , locationStartColumn = startColumn
            , locationEndLine = endLine
            , locationEndColumn = endColumn
            }
    where
        sourceCoordinate context value
            | value <= 0 = Left (modelError context "compiler source positions must be one-based")
            | toInteger value - 1 > toInteger (maxBound :: Word32) =
                Left (modelError context "compiler source position exceeds the protocol range")
            | otherwise = Right (fromIntegral (value - 1))

encodeDiagnosticDocument ::
    DiagnosticProtocolLimits ->
    DiagnosticDocument ->
    Either DiagnosticProtocolError ByteString.ByteString
encodeDiagnosticDocument limits document = do
    validateDocument limits document
    encodedRecords <- traverse (encodeRecord limits) (documentRecords document)
    let payload =
            mconcat (map Builder.word8 magic)
                <> Builder.word16LE protocolVersion
                <> Builder.word16LE 0
                <> countBuilder (length (documentRecords document))
                <> mconcat encodedRecords
        bytes = LazyByteString.toStrict (Builder.toLazyByteString payload)
    if ByteString.length bytes > maximumWireBytes limits
        then Left (limitError "wire byte length" "diagnostic document exceeds configured byte limit")
        else Right bytes

encodeRecord :: DiagnosticProtocolLimits -> DiagnosticRecord -> Either DiagnosticProtocolError Builder.Builder
encodeRecord limits record = do
    arguments <- traverse encodeArgument (recordArguments record)
    primary <- traverse encodeLocation (recordPrimary record)
    related <- traverse encodeRelated (recordRelated record)
    fixes <- traverse (encodeFix limits) (recordFixes record)
    pure
        ( Builder.word8 (stageTag (recordStage record))
            <> Builder.word8 (severityTag (recordSeverity record))
            <> textBuilder (recordCode record)
            <> textBuilder (recordMessage record)
            <> countBuilder (length arguments)
            <> mconcat arguments
            <> Builder.word8 (maybe 0 (const 1) primary)
            <> maybe mempty id primary
            <> countBuilder (length related)
            <> mconcat related
            <> countBuilder (length fixes)
            <> mconcat fixes
        )

encodeArgument :: DiagnosticArgument -> Either DiagnosticProtocolError Builder.Builder
encodeArgument argument =
    pure (textBuilder (argumentName argument) <> textBuilder (argumentValue argument))

encodeLocation :: DiagnosticLocation -> Either DiagnosticProtocolError Builder.Builder
encodeLocation location =
    pure
        ( textBuilder (locationSource location)
            <> Builder.word32LE (locationStartLine location)
            <> Builder.word32LE (locationStartColumn location)
            <> Builder.word32LE (locationEndLine location)
            <> Builder.word32LE (locationEndColumn location)
        )

encodeRelated :: DiagnosticRelatedLocation -> Either DiagnosticProtocolError Builder.Builder
encodeRelated related = do
    location <- encodeLocation (relatedLocation related)
    pure (location <> textBuilder (relatedMessage related))

encodeFix :: DiagnosticProtocolLimits -> DiagnosticFix -> Either DiagnosticProtocolError Builder.Builder
encodeFix _ fix = do
    edits <- traverse encodeEdit (fixEdits fix)
    pure (textBuilder (fixTitle fix) <> countBuilder (length edits) <> mconcat edits)

encodeEdit :: DiagnosticTextEdit -> Either DiagnosticProtocolError Builder.Builder
encodeEdit edit = do
    location <- encodeLocation (editLocation edit)
    pure (location <> textBuilder (editReplacement edit))

textBuilder :: String -> Builder.Builder
textBuilder value =
    countBuilder (length value)
        <> mconcat [Builder.word32LE (fromIntegral (ord character)) | character <- value]

countBuilder :: Int -> Builder.Builder
countBuilder = Builder.word32LE . fromIntegral

stageTag :: DiagnosticStage -> Word8
stageTag = fromIntegral . fromEnum

severityTag :: ProtocolSeverity -> Word8
severityTag = fromIntegral . fromEnum

protocolSeverity :: DiagnosticSeverity -> ProtocolSeverity
protocolSeverity severity = case severity of
    Error -> ProtocolError
    Warning -> ProtocolWarning

validateDocument :: DiagnosticProtocolLimits -> DiagnosticDocument -> Either DiagnosticProtocolError ()
validateDocument limits document = do
    validateCount limits "diagnostic count" (maximumRecords limits) (length (documentRecords document))
    mapM_ (validateRecord limits) (documentRecords document)

validateRecord :: DiagnosticProtocolLimits -> DiagnosticRecord -> Either DiagnosticProtocolError ()
validateRecord limits record = do
    unless
        (validCode (recordCode record))
        (Left (modelError "diagnostic code" "diagnostic code must be 1-64 ASCII uppercase, digit, or hyphen scalars"))
    when
        (null (recordMessage record))
        (Left (modelError "diagnostic message" "diagnostic message is empty"))
    validateText limits "diagnostic code" (recordCode record)
    validateText limits "diagnostic message" (recordMessage record)
    validateCount limits "diagnostic argument count" (maximumArguments limits) (length (recordArguments record))
    let names = map argumentName (recordArguments record)
    when
        (any null names)
        (Left (modelError "diagnostic argument name" "diagnostic argument name is empty"))
    when
        (length names /= length (nub names))
        (Left (modelError "diagnostic argument name" "diagnostic argument names must be unique within one record"))
    mapM_ validateArgument (recordArguments record)
    mapM_ (validateLocation limits "primary location") (recordPrimary record)
    validateCount limits "related location count" (maximumRelatedLocations limits) (length (recordRelated record))
    mapM_ validateRelated (recordRelated record)
    validateCount limits "diagnostic fix count" (maximumFixes limits) (length (recordFixes record))
    mapM_ validateFix (recordFixes record)
    where
        validateArgument argument = do
            validateText limits "diagnostic argument name" (argumentName argument)
            validateText limits "diagnostic argument value" (argumentValue argument)
        validateRelated related = do
            validateLocation limits "related location" (relatedLocation related)
            when
                (null (relatedMessage related))
                (Left (modelError "related location message" "related location message is empty"))
            validateText limits "related location message" (relatedMessage related)
        validateFix fix = do
            when
                (null (fixTitle fix))
                (Left (modelError "diagnostic fix title" "diagnostic fix title is empty"))
            when
                (null (fixEdits fix))
                (Left (modelError "diagnostic fix edits" "diagnostic fix must contain at least one edit"))
            validateText limits "diagnostic fix title" (fixTitle fix)
            validateCount limits "diagnostic edit count" (maximumEditsPerFix limits) (length (fixEdits fix))
            mapM_ validateEdit (fixEdits fix)
        validateEdit edit = do
            validateLocation limits "diagnostic edit" (editLocation edit)
            validateText limits "diagnostic edit replacement" (editReplacement edit)

validateLocation :: DiagnosticProtocolLimits -> String -> DiagnosticLocation -> Either DiagnosticProtocolError ()
validateLocation limits context location = do
    when
        (null (locationSource location))
        (Left (modelError context "diagnostic source identity is empty"))
    validateText limits "diagnostic source" (locationSource location)
    unless
        ( positionPrecedes
            (locationStartLine location, locationStartColumn location)
            (locationEndLine location, locationEndColumn location)
        )
        (Left (modelError context "diagnostic range end precedes its start"))

positionPrecedes :: (Word32, Word32) -> (Word32, Word32) -> Bool
positionPrecedes (leftLine, leftColumn) (rightLine, rightColumn) =
    leftLine < rightLine || (leftLine == rightLine && leftColumn <= rightColumn)

validateText :: DiagnosticProtocolLimits -> String -> String -> Either DiagnosticProtocolError ()
validateText limits context value = do
    validateCount limits context (maximumTextScalars limits) (length value)
    unless
        (all isUnicodeScalar value)
        (Left (modelError context "text contains a non-scalar Unicode value"))

validateCount :: DiagnosticProtocolLimits -> String -> Int -> Int -> Either DiagnosticProtocolError ()
validateCount _ context maximumCount value
    | value < 0 || value > maximumCount || toInteger value > toInteger (maxBound :: Word32) =
        Left (limitError context "collection count exceeds configured limit")
    | otherwise = Right ()

validCode :: String -> Bool
validCode code =
    not (null code)
        && length code <= 64
        && all (\character -> isAsciiUpper character || isDigit character || character == '-') code

isUnicodeScalar :: Char -> Bool
isUnicodeScalar character =
    let value = ord character
     in value <= 0x10ffff && not (value >= 0xd800 && value <= 0xdfff)

modelError :: String -> String -> DiagnosticProtocolError
modelError = DiagnosticProtocolError 0

limitError :: String -> String -> DiagnosticProtocolError
limitError = DiagnosticProtocolError 0

data DecodeState = DecodeState
    { decodeBytes :: ByteString.ByteString
    , decodeOffset :: Int
    , decodeLimits :: DiagnosticProtocolLimits
    }

newtype Decoder value = Decoder
    { runDecoder :: DecodeState -> Either DiagnosticProtocolError (value, DecodeState)
    }

instance Functor Decoder where
    fmap transform parser = Decoder $ \state -> do
        (value, next) <- runDecoder parser state
        pure (transform value, next)

instance Applicative Decoder where
    pure value = Decoder (Right . (value,))
    function <*> argument = Decoder $ \state -> do
        (transform, afterFunction) <- runDecoder function state
        (value, afterArgument) <- runDecoder argument afterFunction
        pure (transform value, afterArgument)

instance Monad Decoder where
    parser >>= continuation = Decoder $ \state -> do
        (value, next) <- runDecoder parser state
        runDecoder (continuation value) next

decodeDiagnosticDocument ::
    DiagnosticProtocolLimits ->
    ByteString.ByteString ->
    Either DiagnosticProtocolError DiagnosticDocument
decodeDiagnosticDocument limits bytes
    | ByteString.length bytes > maximumWireBytes limits =
        Left (limitError "wire byte length" "diagnostic document exceeds configured byte limit")
    | otherwise = do
        (document, state) <- runDecoder decodeDocument (DecodeState bytes 0 limits)
        unless
            (decodeOffset state == ByteString.length bytes)
            (Left (DiagnosticProtocolError (decodeOffset state) "document" "bytes remain after diagnostic document"))
        validateDocument limits document
        pure document

decodeDocument :: Decoder DiagnosticDocument
decodeDocument = do
    observedMagic <- replicateM 4 (byte "magic")
    unlessDecoder (observedMagic == magic) "magic" "input is not a Visual X# diagnostic document"
    version <- word16 "version"
    unlessDecoder (version == protocolVersion) "version" "unsupported diagnostic protocol version"
    flags <- word16 "flags"
    unlessDecoder (flags == 0) "flags" "reserved diagnostic flags must be zero"
    count <- boundedCount "diagnostic count" maximumRecords
    DiagnosticDocument <$> replicateM count decodeRecord

decodeRecord :: Decoder DiagnosticRecord
decodeRecord = do
    stage <- decodeStage
    severity <- decodeSeverity
    code <- text "diagnostic code"
    message <- text "diagnostic message"
    argumentCount <- boundedCount "diagnostic argument count" maximumArguments
    arguments <-
        replicateM argumentCount (DiagnosticArgument <$> text "diagnostic argument name" <*> text "diagnostic argument value")
    hasPrimary <- boolean "primary location presence"
    primary <- if hasPrimary then Just <$> decodeLocation else pure Nothing
    relatedCount <- boundedCount "related location count" maximumRelatedLocations
    related <- replicateM relatedCount (DiagnosticRelatedLocation <$> decodeLocation <*> text "related location message")
    fixCount <- boundedCount "diagnostic fix count" maximumFixes
    fixes <- replicateM fixCount decodeFix
    pure (DiagnosticRecord stage severity code message arguments primary related fixes)

decodeLocation :: Decoder DiagnosticLocation
decodeLocation =
    DiagnosticLocation
        <$> text "diagnostic source"
        <*> word32 "diagnostic start line"
        <*> word32 "diagnostic start column"
        <*> word32 "diagnostic end line"
        <*> word32 "diagnostic end column"

decodeFix :: Decoder DiagnosticFix
decodeFix = do
    title <- text "diagnostic fix title"
    editCount <- boundedCount "diagnostic edit count" maximumEditsPerFix
    edits <- replicateM editCount (DiagnosticTextEdit <$> decodeLocation <*> text "diagnostic edit replacement")
    pure (DiagnosticFix title edits)

decodeStage :: Decoder DiagnosticStage
decodeStage = do
    tag <- byte "diagnostic stage"
    if fromIntegral tag <= fromEnum (maxBound :: DiagnosticStage)
        then pure (toEnum (fromIntegral tag))
        else failDecoder "diagnostic stage" "unknown diagnostic stage tag"

decodeSeverity :: Decoder ProtocolSeverity
decodeSeverity = do
    tag <- byte "diagnostic severity"
    if fromIntegral tag <= fromEnum (maxBound :: ProtocolSeverity)
        then pure (toEnum (fromIntegral tag))
        else failDecoder "diagnostic severity" "unknown diagnostic severity tag"

text :: String -> Decoder String
text context = do
    count <- boundedCount context maximumTextScalars
    values <- replicateM count (word32 context)
    if all validScalar values
        then pure (map (toEnum . fromIntegral) values)
        else failDecoder context "text contains a non-scalar Unicode value"
    where
        validScalar value = value <= 0x10ffff && not (value >= 0xd800 && value <= 0xdfff)

boundedCount :: String -> (DiagnosticProtocolLimits -> Int) -> Decoder Int
boundedCount context selector = do
    value <- word32 context
    limits <- Decoder (\state -> Right (decodeLimits state, state))
    if toInteger value <= toInteger (selector limits)
        then pure (fromIntegral value)
        else failDecoder context "collection count exceeds configured limit"

boolean :: String -> Decoder Bool
boolean context = do
    value <- byte context
    case value of
        0 -> pure False
        1 -> pure True
        _ -> failDecoder context "boolean byte must be zero or one"

byte :: String -> Decoder Word8
byte context = Decoder $ \state ->
    if decodeOffset state >= ByteString.length (decodeBytes state)
        then Left (DiagnosticProtocolError (decodeOffset state) context "input ended before field was complete")
        else
            Right
                ( ByteString.index (decodeBytes state) (decodeOffset state)
                , state {decodeOffset = decodeOffset state + 1}
                )

word16 :: String -> Decoder Word16
word16 context = do
    low <- byte context
    high <- byte context
    pure (fromIntegral low .|. (fromIntegral high `shiftL` 8))

word32 :: String -> Decoder Word32
word32 context = do
    a <- byte context
    b <- byte context
    c <- byte context
    d <- byte context
    pure
        ( fromIntegral a
            .|. (fromIntegral b `shiftL` 8)
            .|. (fromIntegral c `shiftL` 16)
            .|. (fromIntegral d `shiftL` 24)
        )

unlessDecoder :: Bool -> String -> String -> Decoder ()
unlessDecoder condition context message =
    if condition then pure () else failDecoder context message

failDecoder :: String -> String -> Decoder value
failDecoder context message = Decoder $ \state ->
    Left (DiagnosticProtocolError (decodeOffset state) context message)

File: Compiler/Diagnostic/Protocol.cpp
// SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
// SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.1

#include <algorithm>
#include <array>
#include <limits>
#include <string_view>
#include <unordered_set>
#include <utility>

#include "Visual/XSharp/Diagnostic/Protocol.hpp"

namespace Visual::XSharp::Diagnostic
{
    namespace Wire = Artifact::Wire;

    namespace
    {
        constexpr std::array<std::uint8_t, 4U> kMagic{ 'V', 'X', 'D', 'G' };

        [[nodiscard]] auto
        WireLimits(const Limits &limits) -> Wire::Limits
        {
            Wire::Limits result;
            result.maximumWireBytes = limits.maximumWireBytes;
            result.maximumTextScalars = limits.maximumTextScalars;
            result.maximumFunctions = limits.maximumRecords;
            result.maximumParameters = std::max({ limits.maximumArguments,
                                                  limits.maximumRelatedLocations,
                                                  limits.maximumFixes,
                                                  limits.maximumEditsPerFix });
            return result;
        }

        [[nodiscard]] auto
        Failure(ErrorKind kind, std::size_t offset, std::string context, std::string message)
            -> Error
        {
            return Error{ kind, offset, std::move(context), std::move(message) };
        }

        [[nodiscard]] auto
        PositionPrecedes(const Position &left, const Position &right) noexcept -> bool
        {
            return left.line < right.line
                   || (left.line == right.line && left.column <= right.column);
        }

        [[nodiscard]] auto
        ValidStage(Stage stage) noexcept -> bool
        {
            return static_cast<std::uint8_t>(stage)
                   <= static_cast<std::uint8_t>(Stage::LlvmBackend);
        }

        [[nodiscard]] auto
        ValidSeverity(Severity severity) noexcept -> bool
        {
            return static_cast<std::uint8_t>(severity)
                   <= static_cast<std::uint8_t>(Severity::Hint);
        }

        [[nodiscard]] auto
        ValidCode(std::u32string_view code) noexcept -> bool
        {
            if (code.empty() || code.size() > 64U)
                return false;
            return std::ranges::all_of(code, [](char32_t character) {
                return (character >= U'A' && character <= U'Z')
                       || (character >= U'0' && character <= U'9')
                       || character == U'-';
            });
        }

        [[nodiscard]] auto
        ValidateLocation(const Location &location, std::string_view context)
            -> std::optional<Error>
        {
            if (location.source.empty())
                return Failure(ErrorKind::InvalidModel, 0U, std::string(context), "diagnostic source identity is empty");
            if (!PositionPrecedes(location.range.start, location.range.end))
                return Failure(ErrorKind::InvalidModel, 0U, std::string(context), "diagnostic range end precedes its start");
            return std::nullopt;
        }

        [[nodiscard]] auto
        ValidateRecord(const Record &record, const Limits &limits) -> std::optional<Error>
        {
            if (!ValidStage(record.stage))
                return Failure(ErrorKind::InvalidTag, 0U, "diagnostic stage", "diagnostic stage is outside the v1 catalog");
            if (!ValidSeverity(record.severity))
                return Failure(ErrorKind::InvalidTag, 0U, "diagnostic severity", "diagnostic severity is outside the v1 catalog");
            if (!ValidCode(record.code))
                return Failure(ErrorKind::InvalidModel, 0U, "diagnostic code", "diagnostic code must be 1-64 ASCII uppercase, digit, or hyphen scalars");
            if (record.message.empty())
                return Failure(ErrorKind::InvalidModel, 0U, "diagnostic message", "diagnostic message is empty");
            if (record.arguments.size() > limits.maximumArguments)
                return Failure(ErrorKind::LimitExceeded, 0U, "diagnostic argument count", "diagnostic argument count exceeds configured limit");
            if (record.related.size() > limits.maximumRelatedLocations)
                return Failure(ErrorKind::LimitExceeded, 0U, "related location count", "related location count exceeds configured limit");
            if (record.fixes.size() > limits.maximumFixes)
                return Failure(ErrorKind::LimitExceeded, 0U, "diagnostic fix count", "diagnostic fix count exceeds configured limit");

            std::unordered_set<std::u32string> argumentNames;
            argumentNames.reserve(record.arguments.size());
            for (const auto &argument : record.arguments)
            {
                if (argument.name.empty())
                    return Failure(ErrorKind::InvalidModel, 0U, "diagnostic argument name", "diagnostic argument name is empty");
                if (!argumentNames.insert(argument.name).second)
                    return Failure(ErrorKind::InvalidModel, 0U, "diagnostic argument name", "diagnostic argument names must be unique within one record");
            }
            if (record.primary)
                if (auto error = ValidateLocation(*record.primary, "primary location"))
                    return error;
            for (const auto &related : record.related)
            {
                if (related.message.empty())
                    return Failure(ErrorKind::InvalidModel, 0U, "related location message", "related location message is empty");
                if (auto error = ValidateLocation(related.location, "related location"))
                    return error;
            }
            for (const auto &fix : record.fixes)
            {
                if (fix.title.empty())
                    return Failure(ErrorKind::InvalidModel, 0U, "diagnostic fix title", "diagnostic fix title is empty");
                if (fix.edits.empty())
                    return Failure(ErrorKind::InvalidModel, 0U, "diagnostic fix edits", "diagnostic fix must contain at least one edit");
                if (fix.edits.size() > limits.maximumEditsPerFix)
                    return Failure(ErrorKind::LimitExceeded, 0U, "diagnostic edit count", "diagnostic edit count exceeds configured limit");
                for (const auto &edit : fix.edits)
                    if (auto error = ValidateLocation(edit.location, "diagnostic edit"))
                        return error;
            }
            return std::nullopt;
        }

        void
        WriteLocation(Wire::Writer &writer, const Location &location)
        {
            writer.Text(location.source, "diagnostic source");
            writer.U32(location.range.start.line);
            writer.U32(location.range.start.column);
            writer.U32(location.range.end.line);
            writer.U32(location.range.end.column);
        }

        void
        WriteRecord(Wire::Writer &writer, const Record &record, const Limits &limits)
        {
            writer.Byte(static_cast<std::uint8_t>(record.stage));
            writer.Byte(static_cast<std::uint8_t>(record.severity));
            writer.Text(record.code, "diagnostic code");
            writer.Text(record.message, "diagnostic message");

            writer.Count(record.arguments.size(), limits.maximumArguments, "diagnostic argument count");
            for (const auto &argument : record.arguments)
            {
                writer.Text(argument.name, "diagnostic argument name");
                writer.Text(argument.value, "diagnostic argument value");
            }

            writer.Boolean(record.primary.has_value());
            if (record.primary)
                WriteLocation(writer, *record.primary);

            writer.Count(record.related.size(), limits.maximumRelatedLocations, "related location count");
            for (const auto &related : record.related)
            {
                WriteLocation(writer, related.location);
                writer.Text(related.message, "related location message");
            }

            writer.Count(record.fixes.size(), limits.maximumFixes, "diagnostic fix count");
            for (const auto &fix : record.fixes)
            {
                writer.Text(fix.title, "diagnostic fix title");
                writer.Count(fix.edits.size(), limits.maximumEditsPerFix, "diagnostic edit count");
                for (const auto &edit : fix.edits)
                {
                    WriteLocation(writer, edit.location);
                    writer.Text(edit.replacement, "diagnostic edit replacement");
                }
            }
        }

        [[nodiscard]] auto
        ReadLocation(Wire::Reader &reader) -> Location
        {
            Location location;
            location.source = reader.Text("diagnostic source");
            location.range.start.line = reader.U32("diagnostic start line");
            location.range.start.column = reader.U32("diagnostic start column");
            location.range.end.line = reader.U32("diagnostic end line");
            location.range.end.column = reader.U32("diagnostic end column");
            return location;
        }

        [[nodiscard]] auto
        ReadStage(Wire::Reader &reader) -> Stage
        {
            const auto tag = reader.Byte("diagnostic stage");
            if (tag > static_cast<std::uint8_t>(Stage::LlvmBackend))
            {
                reader.Fail(ErrorKind::InvalidTag, "diagnostic stage", "unknown diagnostic stage tag");
                return Stage::Parser;
            }
            return static_cast<Stage>(tag);
        }

        [[nodiscard]] auto
        ReadSeverity(Wire::Reader &reader) -> Severity
        {
            const auto tag = reader.Byte("diagnostic severity");
            if (tag > static_cast<std::uint8_t>(Severity::Hint))
            {
                reader.Fail(ErrorKind::InvalidTag, "diagnostic severity", "unknown diagnostic severity tag");
                return Severity::Error;
            }
            return static_cast<Severity>(tag);
        }

        [[nodiscard]] auto
        ReadRecord(Wire::Reader &reader, const Limits &limits) -> Record
        {
            Record record;
            record.stage = ReadStage(reader);
            record.severity = ReadSeverity(reader);
            record.code = reader.Text("diagnostic code");
            record.message = reader.Text("diagnostic message");

            const auto argumentCount = reader.Count(limits.maximumArguments, "diagnostic argument count");
            record.arguments.reserve(argumentCount);
            for (std::size_t index = 0U; index < argumentCount && !reader.Failure(); ++index)
                record.arguments.push_back(
                    { reader.Text("diagnostic argument name"),
                      reader.Text("diagnostic argument value") });

            if (reader.Boolean("primary location presence"))
                record.primary = ReadLocation(reader);

            const auto relatedCount = reader.Count(limits.maximumRelatedLocations, "related location count");
            record.related.reserve(relatedCount);
            for (std::size_t index = 0U; index < relatedCount && !reader.Failure(); ++index)
            {
                auto location = ReadLocation(reader);
                record.related.push_back(
                    { std::move(location), reader.Text("related location message") });
            }

            const auto fixCount = reader.Count(limits.maximumFixes, "diagnostic fix count");
            record.fixes.reserve(fixCount);
            for (std::size_t fixIndex = 0U; fixIndex < fixCount && !reader.Failure(); ++fixIndex)
            {
                Fix fix;
                fix.title = reader.Text("diagnostic fix title");
                const auto editCount = reader.Count(limits.maximumEditsPerFix, "diagnostic edit count");
                fix.edits.reserve(editCount);
                for (std::size_t editIndex = 0U; editIndex < editCount && !reader.Failure(); ++editIndex)
                {
                    auto location = ReadLocation(reader);
                    fix.edits.push_back(
                        { std::move(location), reader.Text("diagnostic edit replacement") });
                }
                record.fixes.push_back(std::move(fix));
            }
            return record;
        }
    } // namespace

    auto
    Encode(const Document &document, const Limits &limits) -> EncodeResult
    {
        if (document.records.size() > limits.maximumRecords)
            return { {}, Failure(ErrorKind::LimitExceeded, 0U, "diagnostic count", "diagnostic count exceeds configured limit") };
        for (const auto &record : document.records)
            if (auto error = ValidateRecord(record, limits))
                return { {}, std::move(error) };

        auto wireLimits = WireLimits(limits);
        Wire::Writer writer(wireLimits);
        for (const auto byte : kMagic)
            writer.Byte(byte);
        writer.U16(kProtocolVersion);
        writer.U16(0U);
        writer.Count(document.records.size(), limits.maximumRecords, "diagnostic count");
        for (const auto &record : document.records)
            WriteRecord(writer, record, limits);
        if (writer.Failure())
            return { {}, writer.Failure() };
        return { writer.TakeBytes(), std::nullopt };
    }

    auto
    Decode(const std::vector<std::uint8_t> &bytes, const Limits &limits) -> DecodeResult
    {
        if (bytes.size() > limits.maximumWireBytes)
            return { std::nullopt,
                     Failure(ErrorKind::LimitExceeded, 0U, "wire byte length", "diagnostic document exceeds configured byte limit") };

        auto wireLimits = WireLimits(limits);
        Wire::Reader reader(bytes, wireLimits);
        for (const auto expected : kMagic)
            if (reader.Byte("magic") != expected)
            {
                reader.Fail(ErrorKind::InvalidMagic, "magic", "input is not a Visual X# diagnostic document");
                break;
            }
        const auto version = reader.U16("version");
        if (!reader.Failure() && version != kProtocolVersion)
            reader.Fail(ErrorKind::UnsupportedVersion, "version", "unsupported diagnostic protocol version");
        const auto flags = reader.U16("flags");
        if (!reader.Failure() && flags != 0U)
            reader.Fail(ErrorKind::InvalidTag, "flags", "reserved diagnostic flags must be zero");

        Document document;
        const auto count = reader.Count(limits.maximumRecords, "diagnostic count");
        document.records.reserve(count);
        for (std::size_t index = 0U; index < count && !reader.Failure(); ++index)
        {
            auto record = ReadRecord(reader, limits);
            if (!reader.Failure())
            {
                if (auto error = ValidateRecord(record, limits))
                    reader.Fail(error->kind, std::move(error->context), std::move(error->message));
            }
            document.records.push_back(std::move(record));
        }
        if (!reader.Failure() && !reader.AtEnd())
            reader.Fail(ErrorKind::TrailingInput, "document", "bytes remain after diagnostic document");
        if (reader.Failure())
            return { std::nullopt, reader.Failure() };
        return { std::move(document), std::nullopt };
    }
} // namespace Visual::XSharp::Diagnostic
